### PR TITLE
cmake: Toolchain abstraction: Allow for out-of-tree cmake includes

### PR DESCRIPTION
**Pull request by daor-oti**
This allows for inclusion of out-of-tree toolchain cmake files
relating to compiler and linker for both target and generic
toolchains.
The base path used was ZEPHYR_BASE, instead of TOOLCHAIN_ROOT, thus
making it impossible to load the out-of-tree toolchain specific
cmake files.
In addition, the generic toolchain may now specify a generic cmake
file for the linker, similar to the target toolchain linker.

The intent here is to abstract Zephyr's dependence on toolchains,
thus allowing for easier porting to other, perhaps commercial,
toolchains and/or usecases.

No functional change expected.

Part of #16031

### CODE REVIEW

**Review by mped-oticon — APPROVED** [review]
LGTM

**Review by nashif — APPROVED** [review]

**Review by aescolar — APPROVED** [review]

**Comment by aescolar** [comment]
As I understand @SebastianBoe is on holiday for the next 1.5weeks still.
This PR would seem very uncontroversial, and we have an approval from the other cmake files codeowner, so I will proceed to merge it.

**Pull request merged**
